pmd-core/src/main/java/net/sourceforge/pmd/SourceCodeProcessor.java

```diff
@@ -173,6 +173,34 @@ private void processSource(Reader sourceCode, RuleSets ruleSets, RuleContext ctx
         Parser parser = PMD.parserFor(languageVersion, configuration);
 
         Node rootNode = parse(ctx, sourceCode, parser);
+        // basically:
+        // 1. make the union of all stage dependencies of each rule, by language, for the Rulesets
+        // 2. order them by dependency
+        // 3. run them and time them if needed
+
+        // The problem is the first two steps need only be done once.
+        // They're probably costly and if we do this here without changing anything,
+        // they'll be done on each file! Btw currently the "usesDfa" and such are nested loops testing
+        // all rules of all rulesets, but they're run on each file too!
+
+        // Also, the benchmarking framework needs a small refactor. TimedOperationCategory needs to be
+        // made extensible -> probably should be turned to a class with static constants + factory methods
+        // and not an enum.
```

Comment:
actually, I wouldn't do this. The current implementation allows to track under any given category with a label, and the report will take labeled entries and group them together, along with a grand total (this is how rules are tracked). So, we could simply drop the symbol table, dfa and type resolution members from the enum, have a `PREPROCESSING_STAGE` or similar, and have each one track itself under a label (terse name). It will even show how many times each stage was actually run by default.

The current implementation should be flexible enough while keeping uniqueness of entries and order (thanks to them being an enum).